primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,184 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Offchain Storage Lock
+//!
+//! A persistent storage lock with a defined expire time.
+//!
+//! One use case would be collective updates of multiple data items
+//! or append / remove of i.e. sets, vectors which are stored in
+//! the offchain storage DB.
+//!
+//! ## Example:
+//!
+//! ```rust
+//! // in your off-chain worker code
+//!
+//! fn append_to_in_storage_vec<'k, T>(key: &'k [u8], T) where T: Encode {
+//!    let mut lock = StorageLock::new(b"x::lock");
+//!    if let Ok(_guard) =  lock.spin_lock() {
+//!         let acc = StorageValueRef::persistent(key);
+//!         let v: Vec<T> = acc.get::<Vec<T>>().unwrap().unwrap();
+//!         // modify `v` as desired
+//!         acc.set(v);
+//!    } else {
+//!         // the lock duration expired
+//!    }
+//! }
+//! ```
+
+use crate::offchain::storage::StorageValueRef;
+use sp_core::offchain::{Duration, Timestamp};
+use sp_io::offchain;
+
+/// Default expirey duration in milliseconds.
+const STORAGE_LOCK_DEFAULT_EXPIREY_DURATION: u64 = 30_000;
+
+/// Snooze duration before attempting to lock again in ms.
+const STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE: u64 = 100;
+
+/// A persisted guard state.
+///
+/// An in DB persistent mutex for multi access items which are modified
+/// i.e. vecs or sets.
+pub struct StorageLock<'a> {
+	// A storage value ref which defines the DB entry representing the lock.
+	value_ref: StorageValueRef<'a>,
+	// `None` implies it was already released by `fn unlock(..)`
+	locked_until: Option<Timestamp>,
+}
+
+impl<'a> StorageLock<'a> {
+	/// Create a new storage lock with [default expirey duration](Self::STORAGE_LOCK_DEFAULT_EXPIREY_DURATION).
+	pub fn new<'k>(key: &'k [u8]) -> Self
+	where
+		'k: 'a,
+	{
+		Self {
+			value_ref: StorageValueRef::<'a>::persistent(key),
+			locked_until: None,
+		}
+	}
+
+	#[inline]
+	fn try_lock_inner(&mut self, duration: Duration) -> Result<(), Option<Timestamp>> {
+		let now = offchain::timestamp();
+		let expires_at = now.add(duration);
+		let res = self.value_ref.mutate(
+			|s: Option<Option<Timestamp>>| -> Result<Timestamp, Option<Timestamp>> {
+				match s {
+					// no lock set, we can safely acquire it
+					None => Ok(expires_at),
+					// lock is set, but it's old. We can re-acquire it.
+					Some(Some(current_good_until)) if current_good_until < now => Ok(expires_at),
+					// lock is present and is still active
+					Some(Some(current_good_until)) => Err(Some(current_good_until)),
+					_ => Err(None),
+				}
+			},
+		);
+		match res {
+			Ok(Ok(_)) => {
+				self.locked_until = Some(expires_at);
+				Ok(())
+			}
+			Ok(Err(timestamp)) => Err(Some(timestamp)), // failed to set the new value, but could read the current
```

Comment:
I know the API is a bit confusing, but this is not correct. `timestamp` will always be `expires_at` in that case. The `Ok(Err` variant is returned if we failed to set the value (cause some other off-chain worker did that) - so in current case, where all locks put the same expiration it's going to be correct, but if we ever decide to let locks customize the locking value it won't be equal to the value that is in the storage.

What's the use case of returning the lock time anyway? Is the idea to handle the error by sleeping till the required timestamp and then re-trying? I feel like doing `spin_lock` forever should be the wiser choice here.

Reply:
Returning the lock time is used to determine the duration for the next lock attempt.

Reply:
With `trait Lockable` the `fn expired(..)` determines if the lock is expired, which than can take advantage of the hint given by the current timestamp.